src/document.js

```diff
@@ -1097,6 +1114,121 @@ class DocumentMask {
     return this._sortedPaths.length === 0;
   }
 
+  /**
+   * Removes the specified values from a sorted field path array.
+   *
+   * @private
+   * @param {Array.<FieldPath>} input - A sorted array of FieldPaths.
+   * @param {Array.<FieldPath>} values - An array of FieldPaths to remove.
+   */
+  static removeFromSortedInput(input, values) {
+    for (let i = 0; i < input.length; ) {
+      let removed = false;
+
+      for (const fieldPath of values) {
+        if (input[i].isEqual(fieldPath)) {
+          input.splice(i, 1);
+          removed = true;
+          break;
+        }
+      }
+
+      if (!removed) {
+        ++i;
+      }
+    }
+  }
+
+  /**
+   * Removes the field path specified in 'fieldPaths' from this document mask.
+   *
+   * @private
+   * @param {Array.<FieldPath>} fieldPaths An array of FieldPaths.
+   */
+  removeFields(fieldPaths) {
+    DocumentMask.removeFromSortedInput(this._sortedPaths, fieldPaths);
+  }
+
+  /**
+   * Returns whether this document mask contains 'fieldPath'.
+   *
+   * @private
+   * @param {FieldPath} fieldPath The field path to test.
+   * @return {boolean} Whether this document mask contains 'fieldPath'.
+   */
+  contains(fieldPath) {
+    for (const sortedPath of this._sortedPaths) {
+      const cmp = sortedPath.compareTo(fieldPath);
+
+      if (cmp === 0) {
+        return true;
+      } else if (cmp > 0) {
+        return false;
+      }
+    }
+
+    return false;
+  }
+
+  /**
+   * Removes all properties from 'data' that are not contained in this document
+   * mask.
+   *
+   * @private
+   * @param {Object} data - An object to filter.
+   * @return {Object} A shallow copy of the object filtered by this document
+   * mask.
+   */
+  applyTo(data) {
+    /*!
+     * Applies this DocumentMask to 'data' and computes the list of field paths
+     * that were specified in the mask but not present in 'data'.
+     */
+    const applyDocumentMask = data => {
+      const remainingPaths = this._sortedPaths.slice(0);
+
+      const processObject = (currentData, currentPath) => {
+        let result = null;
```

Comment:
Does this mean that `applyTo` can return `null`? Does the final line of the function need to be `return result.filteredData || {}` ?

Reply:
Thinking more about this, maybe either comment about why it can't happen, or add a test that causes it to happen and verify that it's handled.

Reply:
I added a test for this. The function does return null, but since I am only using "for ... in" loops to process this data, nothing blew up. Yay.

Either way, I think we should not rely on this :) I changed this function to return an empty object instead of null. Note that while I am processing the nested objects, I still have to return null as the intermediate state.